lib/einhorn/config_file.rb

```diff
@@ -0,0 +1,182 @@
+require 'yaml'
+
+module Einhorn
+  class ConfigFile
+    def self.check_config_file!
+      return unless (config = load_config!)
+      annotate_error('the top-level') do
+        assert_keys(config, %w{bind})
+      end
+
+      config['bind'].each_with_index do |bind, i|
+        annotate_error("at index #{i} under binds") do
+          assert_keys(config, %w{addr port flags fd}, where)
+          assert_type(config, 'addr', String)
+          assert_type(config, 'port', Integer)
+          assert_type(config, 'flags', Array) if config['flags']
+          assert_type(config, 'fd', Integer) if config['fd']
+        end
+      end
+    end
+
+    def self.assert_type(hash, key, expected)
+      value = hash[key]
+      if !value.kind_of?(expected)
+        raise "Expected #{key} to be of type #{expected}, not #{value.class} (#{value.inspect})"
+      end
+    end
+
+    def self.assert_keys(hash, allowed_keys)
+      if !hash.kind_of?(Hash)
+        raise "Expected a Hash, but got a #{hash.class} (#{hash.inspect}"
+      end
+
+      extra = hash.keys - allowed_keys
+      if extra.length > 0
+        raise "Unrecognized keys: #{extra.inspect}"
+      end
+    end
+
+    def self.annotate_error(where, &blk)
```

Comment:
This feels unnecessarily and confusingly clever. Can we do this more directly instead of rewriting exceptions? (Aren't the strings frozen some percentage of the time anyway?)